File: smsparser/src/main/java/de/adorsys/android/smsparser/SmsTool.java
/*
 * Copyright (C) 2017 adorsys GmbH & Co. KG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.adorsys.android.smsparser;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.v4.content.ContextCompat;

public final class SmsTool {
    public static final int REQUEST_CODE_ASK_PERMISSIONS = 123;

    //private constructor to disable initialization
    private SmsTool() {
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    public static void requestSMSPermission(@NonNull Activity activity) {
        final String permission = Manifest.permission.RECEIVE_SMS;
        int hasSpecificPermission = ContextCompat.checkSelfPermission(activity, permission);
        if (hasSpecificPermission != PackageManager.PERMISSION_GRANTED
                && !activity.shouldShowRequestPermissionRationale(permission)) {
            activity.requestPermissions(new String[]{permission},
                    REQUEST_CODE_ASK_PERMISSIONS);
        }
    }
}

File: smsparser/src/main/java/de/adorsys/android/smsparser/SmsTool.kt
package de.adorsys.android.smsparser

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat

object SmsTool {
    const val REQUEST_CODE_ASK_PERMISSIONS = 123

    @JvmStatic
    @RequiresApi(api = Build.VERSION_CODES.M)
    fun requestSMSPermission(activity: Activity) {
        val permission = Manifest.permission.RECEIVE_SMS
        val hasSpecificPermission = ContextCompat.checkSelfPermission(activity, permission)
        if (hasSpecificPermission != PackageManager.PERMISSION_GRANTED && !activity.shouldShowRequestPermissionRationale(permission)) {
            activity.requestPermissions(arrayOf(permission),
                    REQUEST_CODE_ASK_PERMISSIONS)
        }
    }
}